test/integ/change_password_integ_test.js

```diff
@@ -0,0 +1,394 @@
+                            await checkPassword(TEST_USER, "test");
+                            resolve();
+                        })
+                        .catch((err) => {
+                            reject(err);
+                        });
+                })
+                .catch((err) => {
+                    reject(err);
+                });
+        });
+    });
+    it("should prevent user from changing password if user doesn't exist", () => {
+        return performAsyncTest((resolve, reject) => {
+            performUserLogin()
+                .then(async (agent) => {
+                    await checkPassword(TEST_USER, "test");
+                    usersCollection = db.collection("users");
+                    usersCollection.remove({ username: TEST_USER }, true)
+                        .then((result) => {
+                            agent.post("/change_password?type=json")
+                                .type("form")
+                                .send({
+                                    "oldPassword": "test",
+                                    "newPassword": "something",
+                                    "newPasswordConfirm": "something"
+                                })
+                                .then((res) => {
+                                    assert.equal(res.statusCode, 404);
+                                    checkPassword(TEST_USER, "test")
+                                        .then((res) => {
+                                            assert.fail("Password shouldn't change");
+                                        })
+                                        .catch((err) => {
+                                            assert.isDefined(err);
+                                            resolve();
+                                        });
+                                })
+                                .catch((err) => {
+                                    reject(err);
+                                });
+                        })
+                        .catch((err) => {
+                            reject(err);
+                        });
+                })
+                .catch((err) => {
+                    reject(err);
+                });
+        });
+    });
+    // TODO either delete or figure out a way to mock databaseOps.changeUserPassword on the server
+    it("should not change password if changing password encountered an error", (done) => {
+        assert.fail("Not implemented");
+    });
+    it("should not change password if there was an error retrieving session user info from db", () => {
+        return performAsyncTest((resolve, reject) => {
+            performUserLogin()
+                .then(async (agent) => {
+                    await checkPassword(TEST_USER, "test");
+                    usersCollection = db.collection("users");
+                    usersCollection.deleteMany({});
```

Comment:
Too similar to user doesn't exist test, change it so that user session fails in a different way. Perhaps mock it?

Reply:
Going to leave it out. Will probably require that sessions be mocked, which is not possible right now without some refactoring. Not worth the testing effort since the user doesn't exist test provides a comparable result.